Pagination: add PostProcessOptions

wit221 opened the pull request:
This adds a PostProcessOptions that externally managed paginated endpoints need to return and optionally fill options of. This lets Thunder perform additional operations on the output based on the turned-on flags.

The first flags, added in this PR, are `ApplyTextFilter`, which will apply a text filter on the output, and `SetPageInfo`, which will set the page info based on the output.

wedaly commented on graphql/schemabuilder/pagination.go:
```diff
@@ -149,6 +149,14 @@ type PaginationInfo struct {
+// PostProcessOptions can be returned in a PaginateFieldFunc as an argument that succeeds
+// PaginationInfo. When a PaginationInfo is set, it must be succeeded by a PostProcessOptions.
+// The struct is used to instruct Thunder to perform additional operations on the output
+// in the case of manually paginated resolvers.
```
nit: I feel that the last sentence here is the most important, since it explains what this struct means, so maybe that should be the first sentence in the comment.

wedaly commented on graphql/schemabuilder/pagination.go:
```diff
@@ -149,6 +149,14 @@ type PaginationInfo struct {
 }
 
+// PostProcessOptions can be returned in a PaginateFieldFunc as an argument that succeeds
+// PaginationInfo. When a PaginationInfo is set, it must be succeeded by a PostProcessOptions.
```
nit: I found this description a little confusing ("succeeds" has multiple meanings).   Maybe something like: "This struct should be returned as the second result from a manually-paginated field func"?

wedaly commented on graphql/schemabuilder/pagination.go:
```diff
@@ -664,7 +678,7 @@ func (c *connectionContext) getConnection(ctx context.Context, out []reflect.Val
 	// If the pagination is externally managed, thunder isn't going to handle setting page
 	// information or reducing the edges.
-	if c.IsExternallyManaged() {
+	if c.IsExternallyManaged() && !c.PostProcessOptions.ApplyTextFilter {
```
I imagine if we add more postprocessing options that could affect pagination info in the future we'd have to remember to update this `if` statement.  Maybe it would be more maintainable to have this be calculated in a method on `PostProcessOptions`?  Imagining something like:
```
func (p *PostProcessOptions) SetPaginationInfo() bool {
    return p.ApplyTextFilter
}
```

wit221 replied:
Agreed. Going further, however, I would also decouple the ApplyTextFilter with SetPaginationInfo logic. With the above approach, resolvers that will want to set a paginationInfo without having to apply a text filter will be forced to also apply the text filter.

wedaly commented on graphql/schemabuilder/pagination.go:
```diff
@@ -995,6 +1009,11 @@ func (c *connectionContext) parsePaginatedReturnSignature(m *method) (err error)
 		out = out[1:]
+		if len(out) == 0 || out[0] != reflect.TypeOf(PostProcessOptions{}) {
+			err = fmt.Errorf("%s returns a PaginationInfo not followed by a PaginationPostProcessOptions", c.funcType)
+			return
```
just to confirm my understanding: if we accidentally forget to add this to a manually paginated field, we'd expect the service to fail on startup?  (that's what I'd like to have happen -- it's way scarier if stuff might fail silently)

wit221 replied:
That's correct. This PR also includes a test that verifies just that.

wit221 commented:
@wedaly changelog:
- Improved comments;
- Added a `SetPaginationInfo` option.

Would you care to take another look?

wedaly commented on graphql/schemabuilder/pagination.go:
```diff
@@ -149,6 +149,14 @@ type PaginationInfo struct {
+// This struct should be returned as the second result from a manually-paginated field func.
+type PostProcessOptions struct {
+	ApplyTextFilter   bool
+	SetPaginationInfo bool
```
I'd suggest adding a comment explaining what each of these does.

wit221 commented:
FYI changelog:
- renamed SetPaginationInfo -> SetPageInfo;
- Added comments explaining ApplyTextFilter and SetPageInfo;
- Added tests that test different combinations of the above options: with ApplyTextFilter only, with SetPageInfo only and with both.

CodeBrew28 commented on graphql/schemabuilder/pagination.go:
```diff
@@ -149,6 +149,18 @@ type PaginationInfo struct {
+// This struct should be returned as the second result from a manually-paginated field func.
+type PostProcessOptions struct {
+	// Whether or not Thunder should apply filtering on the output.
+	ApplyTextFilter bool
```
What happens if thunder applies the text filter and filters out too many items?

wit221 replied:
Resolvers should set this option with the possibility in mind that an arbitrary number of items will get filtered out.

A use case is when externally-managed resolvers determine they would like to fall back  to Thunder's filtering and fetch the entire data range (disregarding the `First` parameter). Then, Thunder's potential filtering of data such that the resulting set's length is less than the `First` parameter is just an inherent characteristic of the query and data store - there are just not enough items that fit the provided filter.

CodeBrew28 replied:
Ah, that makes a lot more sense!

CodeBrew28 replied:
Can we maybe add a comment explaining when to use it

wit221 replied:
👍
